pkg/portal/kubeconfig/kubeconfig.go

```diff
@@ -0,0 +1,183 @@
+	baseAccessLog *logrus.Entry
+
+	servingCert      *x509.Certificate
+	elevatedGroupIDs []string
+
+	dbOpenShiftClusters database.OpenShiftClusters
+	dbPortal            database.Portal
+
+	dialer      proxy.Dialer
+	clientCache clientcache.ClientCache
+
+	newToken func() string
+}
+
+func New(baseLog *logrus.Entry,
+	baseAccessLog *logrus.Entry,
+	servingCert *x509.Certificate,
+	elevatedGroupIDs []string,
+	dbOpenShiftClusters database.OpenShiftClusters,
+	dbPortal database.Portal,
+	dialer proxy.Dialer,
+	aadAuthenticatedRouter,
+	unauthenticatedRouter *mux.Router) *kubeconfig {
+	k := &kubeconfig{
+		log:           baseLog,
+		baseAccessLog: baseAccessLog,
+
+		servingCert:      servingCert,
+		elevatedGroupIDs: elevatedGroupIDs,
+
+		dbOpenShiftClusters: dbOpenShiftClusters,
+		dbPortal:            dbPortal,
+
+		dialer:      dialer,
+		clientCache: clientcache.New(time.Hour),
+
+		newToken: func() string { return uuid.NewV4().String() },
+	}
+
+	rp := &httputil.ReverseProxy{
+		Director:  k.director,
+		Transport: roundtripper.RoundTripperFunc(k.roundTripper),
+		ErrorLog:  log.New(k.log.Writer(), "", 0),
+	}
+
+	aadAuthenticatedRouter.NewRoute().Methods(http.MethodPost).Path("/subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}/providers/microsoft.redhatopenshift/openshiftclusters/{resourceName}/kubeconfig/new").HandlerFunc(k.new)
+
+	bearerAuthenticatedRouter := unauthenticatedRouter.NewRoute().Subrouter()
+	bearerAuthenticatedRouter.Use(middleware.Bearer(k.dbPortal))
+	bearerAuthenticatedRouter.Use(middleware.Log(k.baseAccessLog))
+
+	bearerAuthenticatedRouter.PathPrefix("/subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}/providers/microsoft.redhatopenshift/openshiftclusters/{resourceName}/kubeconfig/proxy/").Handler(rp)
+
+	return k
+}
+
+// new creates a new PortalDocument allowing kubeconfig access to a cluster for
+// 6 hours and returns a kubeconfig with the temporary credentials
+func (k *kubeconfig) new(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	resourceID := strings.Join(strings.Split(r.URL.Path, "/")[:9], "/")
+	if !validate.RxClusterID.MatchString(resourceID) {
+		http.Error(w, fmt.Sprintf("invalid resourceId %q", resourceID), http.StatusBadRequest)
+		return
+	}
+
+	elevated := middleware.GroupsIntersect(k.elevatedGroupIDs, ctx.Value(middleware.ContextKeyGroups).([]string))
+
+	token := k.newToken()
+
+	portalDoc := &api.PortalDocument{
+		ID:  token,
+		TTL: int(kubeconfigNewTimeout / time.Second),
+		Portal: &api.Portal{
+			Username: ctx.Value(middleware.ContextKeyUsername).(string),
+			ID:       resourceID,
+			Kubeconfig: &api.Kubeconfig{
+				Elevated: elevated,
+			},
+		},
+	}
+
+	_, err := k.dbPortal.Create(ctx, portalDoc)
+	if err != nil {
+		k.internalServerError(w, err)
+		return
+	}
+
+	b, err := k.makeKubeconfig("https://"+r.Host+resourceID+"/kubeconfig/proxy", token)
+	if err != nil {
+		k.internalServerError(w, err)
+		return
+	}
```

Comment:
If this fails, does not means we gonna be leaking portal documents?

Reply:
No: there is a TTL set on the Portal collection so they will auto-expire.

Reply:
ack